wip-trampoline.md

```diff
@@ -0,0 +1,780 @@
+will even provide a bigger anonymity set).
+
+Nodes that keep track of the whole network should advertise support for `trampoline_routing`. This is an opportunity
+for them to earn more fees than with the default onion routing.
+
+A payer selects a few trampoline nodes and builds a corresponding trampoline onion. It then embeds that trampoline
+onion in the last `hop_payload` of an `onion_packet` destined to the first trampoline node. Computing routes between
+trampoline nodes is deferred to the trampoline nodes themselves.
+
+Merchants may include a few trampoline nodes that are close to them in their invoice to help payers select a good last
+trampoline node. This is optional and payers aren't required to use this routing hint.
+
+## Table of Contents
+
+* [Features](#features)
+* [Packet Structure](#packet-structure)
+  * [Trampoline Onion](#trampoline-onion)
+  * [Paying via trampoline nodes](#paying-via-trampoline-nodes)
+  * [Paying to non-trampoline nodes](#paying-to-non-trampoline-nodes)
+* [Routing Gossip](#routing-gossip)
+  * [Fees and CLTV estimation](#fees-and-cltv-estimation)
+  * [The `node_update` Message](#the-node_update-message)
+  * [Filtering gossip messages](#filtering-gossip-messages)
+* [Fees and CLTV requirements](#fees-and-cltv-requirements)
+* [Appendix A: Examples](#appendix-a-examples)
+  * [Merchant supporting trampoline payments](#merchant-supporting-trampoline-payments)
+  * [Merchant without trampoline support](#merchant-without-trampoline-support)
+* [Appendix B: Trampoline and Rendez-Vous](#appendix-b-trampoline-and-rendez-vous)
+* [Appendix C: Trampoline and MPP](#appendix-c-trampoline-and-mpp)
+* [References](#references)
+
+## Features
+
+Trampoline routing uses the following `globalfeatures` flags:
+
+| Bits  | Name                 | Description                                    | Link |
+|-------|----------------------|------------------------------------------------|------|
+| 10/11 | `trampoline_routing` | This node requires/supports trampoline routing | TBD  |
+
+Feature flags are assigned in pairs, but the mandatory value may not be applicable here.
+
+Gossip messages filtering uses the following `localfeatures` flags:
+
+| Bits  | Name                 | Description                                        | Link |
+|-------|----------------------|----------------------------------------------------|------|
+| 12/13 | `gossip_filters`     | This node requires/supports gossip message filters | TBD  |
+
+## Packet Structure
+
+### Trampoline Onion
+
+The trampoline onion is a fixed-size tlv packet with the following structure:
+
+1. type: 14 (`trampoline_onion_packet`)
+2. data:
+   * [`byte`:`version`]
+   * [`point`:`public_key`]
+   * [`400*byte`:`hop_payloads`]
+   * [`32*byte`:`hmac`]
+
+This results in a `466`-bytes fixed-size tlv where the value has exactly the same format as the `onion_packet` with
```

Comment:
With a fixed `466`-bytes fixed-sizes is maximum size implicitly 120 hops (20 * 6) ? I think not given you have variable size payload, but even variable there is a minimum size to describe meaningfully HTLC routed..

Reply:
I would need to re-do the maths, but I think it's less than that.
With the trampoline onion inside a normal onion, the normal onion route can be at most ~15 hops.
And the number of trampoline hops will be at most 4 or 5 (would need to double-check).
So we would have maximum 60-hops route. And an attacker wouldn't control the route taken between trampoline hops so in reality the route will likely be less than that (this shouldn't increase too much the effectiveness of a capacity attack).

Reply:
Ah you right for the math didn't subtract the trampoline onion, maybe in future versions, sending node should be able to pick trampoline onion size as a parameter, and so increase number of trampolines (privacy would be a good reason)

Reply:
Yeah I'm trying to balance privacy (where long routes are better) with capacity attacks (where long routes are bad). This is an inherent trade-off with today's lightning and it's a real pain ;)